wordpress-mobile/WordPress-iOS
# Jetpack Migration: Do not direct to WordPress if there are prior migration errors

Refs: pcdRpT-1sp-p2#comment-2466, pcdRpT-1sp-p2#comment-2478

As titled, this PR stores the error produced from `ContentMigrationCoordinator` to the App Group. The purpose is for Jetpack to use this information to determine if it should present the sign-in flow or the "Open WordPress" flow.

> **Note**: As a reminder, we show the Load WordPress flow for users that opens the Jetpack app while having the WordPress but have never opened it at least once (to trigger the export process).

The "Open WordPress" flow is now shown when:
- The import process failed,
- The user has a compatible WordPress app installed,
- There is no exported data in the App Group space,
- The WordPress app has no previous export failures.

To enter this flow, we want to be sure that **there is no data to import because the user has never triggered the export process and not because the WP app has previously tried and failed.** Any condition other than the above should bring the user to the sign-in flow.

## To test

As a reminder, here are three possible states when the Jetpack app is opened:

🔐 Sign-in flow | 🟢 Normal migration | 🔁 Open WordPress
-|-|-
![Splash](https://user-images.githubusercontent.com/1299411/208149033-459f3bb3-bcfe-410f-a33c-80627ae540f0.png) | ![Welcome](https://user-images.githubusercontent.com/1299411/208149029-1742fd9c-43ce-4f38-beac-2cb6263dafd6.png) | ![Welcome - JP entry point](https://user-images.githubusercontent.com/1299411/208149022-0d76f149-062a-478c-bc8e-51e7c1138e8d.png)

Next, refer to the table below to see possible setups and variations, along with the expected flow when Jetpack is opened. To test, please follow the conditions in columns 2-4, and verify that the expected flow in the last column is correct.

No. | WordPress app | Account state | Scenario | Expected flow in Jetpack
-|-|-|-|-
1 | Not installed | n/a | n/a | 🔐 Sign-in flow
2 | Incompatible[^4] | Any | Any | 🔐 Sign-in flow
3 | Compatible | Logged out | Never opened | 🔁 Open WordPress
4 | Compatible | Logged out | Opened at least once | 🔐 Sign-in flow[^2]
5 | Compatible | Logged out | Log in > then tap the Jetpack banner | 🟢 Normal migration
6 | Compatible | Logged in | Never opened[^1] | 🔁 Open WordPress
7 | Compatible | Logged in | Opened at least once | 🟢 Normal migration
8 | Compatible | Logged in | Tap the Jetpack banner | 🟢 Normal migration
9 | Compatible | Logged in | Airplane mode > create local draft > tap the Jetpack banner | 🔐 Sign-in flow[^2]
10 | Compatible | Logged in | Log out | 🔐 Sign-in flow[^2]

To be clear, this PR updates the expected flow from **🔁 Open WordPress** to **🔐 Sign-in flow** for rows 4, 9, and 10. However, I'm listing all the other pathways (that I can think of) to ensure they're not impacted by this change.

---

💡 **Pro tip:** If you're planning to test by compiling directly to your device or simulator, it can be tricky to test the `Never opened` scenario. Here's how you can configure Xcode to not launch the app immediately after it finishes building:

1. Click on the app scheme (WordPress or Jetpack), then Edit Scheme...
2. Select the Run tab.
3. In the Launch section, select the "Wait for the executable to be launched" option.

---

## Regression Notes
1. Potential unintended areas of impact
Some migration entry points may be affected. Let's hope not 😬 .

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manually tested the changes.

3. What automated tests I added (or what prevented me from doing so)
Added unit tests to cover additional logic in `ContentMigrationCoordinator`.

PR submission checklist:


[^1]: This is an app update case. From a previous WordPress app, log in to any account, and update the app to the latest version. But DO NOT open it after updating.
[^2]: Previously, this would be 🔁 Open WordPress.
[^4]: This would be version 21.2 and below.
